site/book/05-developing-functions/00.md

```diff
@@ -1,12 +1,78 @@
+languages, frameworks and runtimes. In this chapter, we are going to look at
+different approaches to developing functions.
 
-We are also going to take a closer look at the standardized API between the
-orchestrator (i.e. kpt CLI) and the function containers. Understanding this API
-enables you to have a deeper understanding of how things work under the hood. It
-also enables you to create your own toolchain for function development if you so
-desire.
+?> Before you start developing your custom function, check out the
+[Functions Catalog](https://catalog.kpt.dev/ ":target=_self") in case there is
+an existing function that meets your needs. This is an ever-growing catalog of
+functions that we consider to be generally useful to many users. If your use
+case fits that description, please [open a feature request][fr] for adding a
+function to the catalog.
 
-[ch2]: /book/02-concepts/03-functions
+## Approaches
+
+### Creating Custom Images
+
+With this approach you create a custom container image which can execute
+programs in an arbitrary language or encapsulate existing tools as long as it
+satisfies the KRM Functions Specification we will see later in this chapter.
+
+To simplify development, we provide frameworks for developing functions in Go
+and Typescript. We will cover these later in this chapter.
+
+### Authoring Executable Configuration
+
+Instead of developing a custom image, you can use an existing function image
+containing a language interpreter, and provide your business logic in a KRM
+resource. This is referred to _executable configuration_. We will see two
+examples of executable configuration pattern in this chapter.
+
+Although using executable configuration saves some time initially, it can become
+an anti-pattern if it grows in complexity. We recommend limiting their use to:
+
+- Small amount of logic (< 20 lines)
+- You do not forsee this logic growing in complexity in the future
+
+Otherwise, you are better of developing functions in a general-purpose language
+where you can take advantage of proper abstractions and language features,
+better testing, rich IDE experience, and existing libraries.
+
+## Function Properties
+
+As you think about how to formulate your function, keep in mind the following
+desired properties:
+
+### Deterministic
+
+Executing a function with the same input should produce the same output. For
+example, a function that annotates a resource with the current timestamp is not
+deterministic.
+
+Note that input to the function includes both `input items` and the
+`functionConfig`:
+
+![img](/static/images/func.svg)
+
+### Idempotent
+
+Executing a function a second time should not produce any change. For example, a
+function that increments value of the `replicas` field by 1 is not idempotent.
+Instead, the function could take the desired value of the `replicas` field as
+input.
+
+This property enables in-place edits to work and is analogous to the
```

Comment:
Not sure if we have been using `in-place` term. We can probably substitute this one with "manual edits".

Reply:
There are 7 references to `in-place` in book and reference docs. It should a be a concept we introduce to to users.